packages/mobile/src/walletConnect/walletConnect.ts

```diff
@@ -52,3 +57,21 @@ export function isWalletConnectDeepLink(deepLink: string) {
     decodeURIComponent(deepLink).startsWith(prefix)
   )
 }
+
+export function handleWalletConnectLoadingWithTimeout(
+  routeParams: StackParamList[Screens.WalletConnectLoading]
+) {
+  navigate(Screens.WalletConnectLoading, routeParams)
+
+  setTimeout(async () => {
+    if (await isScreenOnForeground(Screens.WalletConnectLoading)) {
+      ValoraAnalytics.track(WalletConnectEvents.wc_pairing_error, {
+        error: 'timed out while waiting for a session',
+      })
+      navigate(Screens.WalletConnectResult, {
+        title: i18n.t('timeoutTitle'),
+        subtitle: i18n.t('timeoutSubtitle'),
+      })
+    }
+  }, CONNECTION_TIMEOUT)
```

Comment:
This works, but is not very saga-ish and can't be tested as easily.

We could use a `race` with a `delay`. There's a similar example in https://github.com/valora-inc/wallet/blob/a91f80c89efcad47a5cb189e820d11942622b884/packages/mobile/src/fiatExchanges/saga.ts#L70-L77

Reply:
you're right, the setTimeout wasn't very saga-ish...i was thinking of this function more like an async helper, but indeed it is a bit confusing to have this living with other sagas if it's not a saga too.

i ended up doing this very differently, i'm curious for your thoughts. i created a `initialiseWalletConnectWithLoading` generator function that will race `initialiseWalletConnect` with the timeout - i think this makes more sense

Reply:
(wanted to clarify here, that i did try to directly replace the `setTimeout` with a `race` + `delay` but it didn't quite work the same, it seemed the `race` was a blocking rather than async task...maybe i wasn't doing it right)